framework/src/aifw/aifw_csv_reader.c

```diff
@@ -0,0 +1,263 @@
+	}
+}
+
+static void fillValue(void *buffer, CSV_VALUE_DATA_TYPE_E type, uint16_t index, char *value)
+{
+	switch (type) {
+	case INT8:
+		*((int8_t *)buffer + index) = (int8_t)atoi(value);
+		break;
+	case UINT8:
+		*((uint8_t *)buffer + index) = (uint8_t)atoi(value);
+		break;
+	case INT16:
+		*((int16_t *)buffer + index) = (int16_t)atoi(value);
+		break;
+	case INT32:
+		*((int32_t *)buffer + index) = atoi(value);
+		break;
+	case FLOAT32:
+		*((float *)buffer + index) = atof(value);
+		break;
+	default:
+		break;
+	}
+}
+
+static char *skipSpace(char *ptr)
+{
+	while (*ptr && isspace((int)*ptr)) {
+		ptr++;
+	}
+	return ptr;
+}
+
+static uint16_t calculateColumnCount(FILE *handle)
+{
+	int columnCount = 0;
+	int character = fgetc(handle);
+	if (character == EOF || character == '\n' || character == '\r' || character == '\0') {
+		return columnCount;
+	}
+	while (character != EOF && character != '\n' && character != '\r' && character != '\0') {
+		if (character == ',') {
+			columnCount++;
+		}
+		character = fgetc(handle);
+	}
+	columnCount++;
+	return columnCount;
+}
+
+AIFW_RESULT csvInit(void **csvHandle, const char *filename, CSV_VALUE_DATA_TYPE_E dataType, bool hasHeader)
+{
+	if (!csvHandle) {
+		AIFW_LOGE("Double pointer to CSV Handle is NULL.");
+		return AIFW_INVALID_ARG;		
+	}
+	if (!filename) {
+		AIFW_LOGE("file name is null");
+		return AIFW_INVALID_ARG;
+	}
+	if (dataType < INT8 || dataType > FLOAT32) {
+		AIFW_LOGE("Input type is invalid");
+		return AIFW_INVALID_ARG;
+	}
+	*csvHandle = malloc(sizeof(CSVHandle));
+	if (!*csvHandle) {
+		AIFW_LOGE("Memory allocation failed for csv handle");
+		return AIFW_NO_MEM;
+	}
+	CSVHandle *handle = (CSVHandle *)(*csvHandle);
+	handle->columnBuffer = NULL;
+	handle->lineBuffer = NULL;
+	handle->fileHandle = csvOpen(filename);
+	if (!handle->fileHandle) {
+		AIFW_LOGE("file open failed errno: %d", errno);
+		clearMemory(csvHandle);
+		return AIFW_ERROR_FILE_ACCESS;
+	}
+	handle->columnCount = calculateColumnCount(handle->fileHandle);
+	if (handle->columnCount == 0) {
+		AIFW_LOGE("Total column count in csv is 0.");
+		csvDeinit(csvHandle);
+		return AIFW_ERROR;
+	}
+	handle->columnDataType = dataType;
+	GET_MAX_CHARACTERS_PER_VALUE(handle->maxCharPerColumn, dataType);
+	handle->lineCounter = 0;
+	AIFW_LOGV("Total columns count: %d", handle->columnCount);
+	if (!hasHeader && fseek(handle->fileHandle, 0, SEEK_SET)) {
+		AIFW_LOGE("Seeking file pointer to start of file failed.");
+		csvDeinit(csvHandle);
+		return AIFW_ERROR;
+	}
+	handle->columnBuffer = (char *)malloc(handle->maxCharPerColumn * sizeof(char));
+	if (!handle->columnBuffer) {
+		AIFW_LOGE("Memory allocation failed for column buffer");
+		csvDeinit(csvHandle);
+		return AIFW_NO_MEM;
+	}
+	handle->lineBuffer = (char *)malloc(handle->columnCount * handle->maxCharPerColumn * sizeof(char));
+	if (!handle->lineBuffer) {
+		AIFW_LOGE("Memory allocation failed for line buffer");
+		csvDeinit(csvHandle);
+		return AIFW_NO_MEM;
```

Comment:
same as previous
need to check `handle->columnbuffer`, `csv_handle`

Reply:
Here, in case of memory allocation error, we call csvDeinit() which internally clears memory allocated to csv_handle and handle->columnbuffer. So, there won't be any memory leak.